dali/operators/math/expressions/broadcasting.cc

```diff
@@ -0,0 +1,242 @@
+// Copyright (c) 2022, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/math/expressions/broadcasting.h"
+#include "dali/operators/math/expressions/arithmetic_meta.h"
+
+namespace dali {
+
+void CheckNumSamples(span<const TensorListShape<>> shapes) {
+  DALI_ENFORCE(shapes.size() >= 1);
+  int nsamples = shapes[0].num_samples();
+  for (int i = 1; i < shapes.size(); i++) {
+    DALI_ENFORCE(nsamples == shapes[i].num_samples());
+  }
+}
+
+template <typename Shape>
+int BroadcastNdimImpl(span<Shape> shapes) {
+  DALI_ENFORCE(shapes.size() >= 1);
+  int ndim = shapes[0].size();
```

Comment:
I think both TensorShape and TensorListShape have the sample_dim()
```suggestion
  int ndim = shapes[0].sample_dim();
```